:bug: Add request, latency, and workqueue metrics to front proxy

csams opened the pull request:
Signed-off-by: Christopher Sams <csams@redhat.com>

## Summary
Add front proxy specific metrics to enable accurate SLO tracking across shards even in the case of a shard being down.

Fixes #2301

stevekuznetsov commented on pkg/proxy/metrics/metrics.go:
```diff
@@ -0,0 +1,117 @@
+			// This metric is used for verifying api call latencies SLO,
+			// as well as tracking regressions in this aspects.
+			// Thus we customize buckets significantly, to empower both usecases.
+			Buckets: []float64{0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3,
```
Kube provides helpers for these, let's use those if we can. See the authorization cache sync duration metrics we have.

csams replied:
Ack - should I use the granularity used in the auth cache or keep what I lifted from https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/metrics/metrics.go#L101-L102?

csams replied:
I lifted the wrong ones anyway... :grimacing: I'll change it to use the helper.

stevekuznetsov commented on pkg/proxy/metrics/metrics.go:
```diff
@@ -0,0 +1,117 @@
+
+		start := time.Now()
+		delegate.ServeHTTP(sc, r)
+		elapsedSeconds := time.Since(start).Seconds()
```
Depending on how these get layered, you're going to count the time taken to observe metrics at the lower level in this. Might want to think of a factoring that only ends up adding one handler into the chain to avoid that, and any duplicate processing we're going to do (e.g. extracting cluster name, shards, etc from the request)

csams replied:
> Depending on how these get layered, you're going to count the time taken to observe metrics at the lower level in this

I think that's okay unless you feel strongly otherwise. "One wrapper per metric" seems to be a pattern [prometheus uses](https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/instrument_server.go).

> and any duplicate processing we're going to do (e.g. extracting cluster name, shards, etc from the request)

I plan to open a followup PR that will extract the shard info near the top of the chain and attach it to the context so metrics and the shard reverse proxies can use it.

stevekuznetsov commented on pkg/proxy/metrics/metrics.go:
```diff
@@ -0,0 +1,117 @@
+
+		start := time.Now()
+		delegate.ServeHTTP(sc, r)
+		elapsedSeconds := time.Since(start).Seconds()
```
Depending on how these get layered, you're going to count the time taken to observe metrics at the lower level in this. Might want to think of a factoring that only ends up adding one handler into the chain to avoid that, and any duplicate processing we're going to do (e.g. extracting cluster name, shards, etc from the request)

stevekuznetsov commented on pkg/proxy/metrics/metrics.go:
```diff
@@ -0,0 +1,117 @@
+var (
+	requestCounter = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "proxy_request_total",
```
nit: naming convention would suggest `proxy_request_count` - but we actually get the counts for free from the histogram below, so we don't need this metric.

csams replied:
Sure, I can change it if you like. I just slightly modified https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/metrics/metrics.go#L75-L83 and kept the convention there.

stevekuznetsov replied:
With the histogram it's 100% superfluous.

csams replied:
That's fair. We can just use the +Inf bucket. I'll update the PR.

stevekuznetsov commented on pkg/proxy/proxy.go:
```diff
@@ -21,18 +21,18 @@ import (
 
 func newTransport(clientCert, clientKeyFile, caFile string) (*http.Transport, error) {
-	caCert, err := ioutil.ReadFile(caFile)
+	caCert, err := os.ReadFile(caFile)
```
Is this a Go 1.19 thing? Is it valid under 1.18?

s-urbaniak replied:
https://pkg.go.dev/io/ioutil

> Deprecated: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details.

should be valid up until 1.16

stevekuznetsov replied:
ACK

csams replied:
I'm using 1.19, but it should be valid under 1.18. ioutil has been deprecated since 1.16 according to the docs https://pkg.go.dev/io/ioutil

stevekuznetsov commented on pkg/proxy/metrics/metrics.go:
```diff
@@ -0,0 +1,117 @@
+)
+
+// used in the middleware below to capture http status code.
+type statusCatcher struct {
```
Doing this to the round-tripper on the outgoing requests would not require us to do injection like this, right?

stevekuznetsov replied:
Although that might not measure the same latency?

stevekuznetsov replied:
Although that might not measure the same latency?

s-urbaniak commented on pkg/proxy/metrics/metrics.go:
```diff
@@ -0,0 +1,117 @@
+
+// WithLatencyTracking tracks the number of seconds it took the wrapped handler
+// to complete.
+func WithLatencyTracking(delegate http.Handler) http.Handler {
```
you can use https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/instrument_server.go#L84

csams replied:
> // The ObserverVec must have valid metric and label names and must have zero,
// one, or two non-const non-curried labels. For those, the only allowed label
// names are "code" and "method". The function panics otherwise.

I want to come back later and add a shard label.

s-urbaniak replied:
sure, that's still possible with

```
promhttp.InstrumentHandlerDuration(
		durationMetric.MustCurryWith(prometheus.Labels{"shard": "foo"}),
		promhttp.InstrumentHandlerCounter(counterMetric, handler),
	)
```

csams replied:
ahhh thanks! I'll look into using it.

s-urbaniak commented on pkg/proxy/metrics/metrics.go:
```diff
@@ -0,0 +1,117 @@
+
+// WithRequestCounter counts the number of requests that have passed through
+// the handler it wraps.
+func WithRequestCounter(delegate http.Handler) http.Handler {
```
you can use https://github.com/brancz/prometheus-example-app/blob/1ac20239640c6280cbdafc6e9c26b62814c89c97/vendor/github.com/prometheus/client_golang/prometheus/promhttp/instrument_server.go#L94

s-urbaniak commented on pkg/proxy/metrics/metrics.go:
```diff
@@ -0,0 +1,117 @@
+
+// WithRequestCounter counts the number of requests that have passed through
+// the handler it wraps.
+func WithRequestCounter(delegate http.Handler) http.Handler {
```
https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/instrument_server.go#L135

csams replied:
> // The ObserverVec must have valid metric and label names and must have zero,
// one, or two non-const non-curried labels. For those, the only allowed label
// names are "code" and "method". The function panics otherwise.

I want to come back later and add a shard label.

stevekuznetsov replied:
You could just determine the shard at runtime, curry that into the metric, create the delegating handler at runtime and use it?  I'd hate to see us reimplement something that must have been solved as a library with battle-testing behind it :|

stevekuznetsov replied:
Oops, didn't notice the comment here https://github.com/kcp-dev/kcp/pull/2302#discussion_r1013368578 which is the same as mine. Sorry

ncdc commented on pkg/proxy/index/index_controller.go:
```diff
@@ -30,6 +30,8 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
+	_ "k8s.io/component-base/metrics/prometheus/clientgo/leaderelection"
```
I'd probably put these in maybe server.go?

ncdc commented on pkg/proxy/mapping.go:
```diff
@@ -19,11 +19,12 @@ package proxy
 	"net/url"
+	"os"
 
+	"k8s.io/component-base/metrics/legacyregistry"
```
Doesn't have to be solved in this PR if this is good enough to get things going forward, but should we be using metrics/registry instead of metrics/legacyregistry?

ncdc replied:
Let's file an issue to figure out what to do. Being consistent for now is 👍

ncdc commented on pkg/proxy/mapping.go:
```diff
@@ -58,18 +59,18 @@ func NewHandler(ctx context.Context, o *proxyoptions.Options, index index.Index)
 
 	mux := http.NewServeMux()
 
-	// TODO: implement proper readyz handler
```
I would leave this in

ncdc commented on pkg/proxy/mapping.go:
```diff
@@ -58,18 +59,18 @@ func NewHandler(ctx context.Context, o *proxyoptions.Options, index index.Index)
 		w.WriteHeader(http.StatusOK)
 	}))
 
-	// TODO: implement proper livez handler
```
I would leave this in

hardys commented:
@csams this lgtm, and tested locally - one question, do you think we should also add some `proxy_request_total` metrics similar to `apiserver_request_total` ?  

I imagine such metrics will be needed to support e.g an SLO that the service returns above a certain percentage of non-error responses (and will be useful for load-testing), I guess we can add as a follow up :) 

Also do you want to squash the fixup commits here or should we have tide squash on merge?

csams commented:
@hardys I added a request total originally to be consistent with the API server metrics, but @stevekuznetsov preferred to not duplicate the *_count or +Inf metrics that we get already with the histogram, so I removed it.

I'll squash the commits.

hardys commented:
> @hardys I added a request total originally to be consistent with the API server metrics, but @stevekuznetsov preferred to not duplicate the *_count or +Inf metrics that we get already with the histogram, so I removed it.

Ah thanks, I missed that resolved discussion, I'll use the `+Inf` bucket instead :+1:

hardys commented:
/lgtm

csams commented:
/retest
